app/src/main/java/protect/budgetwatch/PasswordManager.java

```diff
@@ -0,0 +1,98 @@
+package protect.budgetwatch;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+
+public class PasswordManager {
+
+    private final String PasswordKey = "password";
+    private final int Iterations = 100;
+    private final int KeyLength = 128;
+
+    private SharedPreferences _prefs;
+
+    PasswordManager(Context context) {
+        _prefs = context.getSharedPreferences("protect.budgetwatch", Context.MODE_PRIVATE);
+    }
+
+    private String getStoredPasswordHash() {
+        return _prefs.getString(PasswordKey, null);
+    }
+
+    public void setPassword(String password) {
+        try {
+            // Generate Salt
+            SecureRandom sr = SecureRandom.getInstance("SHA1PRNG");
+            byte[] salt = new byte[KeyLength / 8];
+            sr.nextBytes(salt);
+            // Generate Hash
+            byte[] hash = generateHash(password, salt);
+            // Write to SharedPreferences
+            _prefs.edit().putString(PasswordKey, toHex(salt) + ":" + toHex(hash)).commit();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    void clearPassword() {
+        _prefs.edit().remove(PasswordKey).commit();
+    }
+
+    boolean isPasswordEnabled() {
+        return getStoredPasswordHash() != null;
+    }
+
+    boolean isPasswordCorrect(String password) {
+        String storedPassword = getStoredPasswordHash();
+        if (storedPassword == null) return false;
+
+        String[] parts = storedPassword.split(":");
+        byte[] salt = fromHex(parts[0]);
+        byte[] hash = fromHex(parts[1]);
+
+        byte[] passwordHash = generateHash(password, salt);
+
+        int diff = hash.length ^ passwordHash.length;
+        for (int i = 0; i < hash.length && i < passwordHash.length; i++) {
+            diff |= hash[i] ^ passwordHash[i];
+        }
+        return diff == 0;
+    }
+
+    private byte[] generateHash(String password, byte[] salt) {
+        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, Iterations, KeyLength);
+        try {
+            SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
+            return skf.generateSecret(spec).getEncoded();
+        } catch (Exception e) {
+            e.printStackTrace();
+            return new byte[0];
+        }
+    }
+
+    private String toHex(byte[] array) {
```

Comment:
Would something like Guava's BaseEncoding be useful?

https://google.github.io/guava/releases/16.0/api/docs/com/google/common/io/BaseEncoding.html

Reply:
Changed.